### DISCUSSION TITLE
Unable to Bypass Face Verification for Singpass Testing

### DISCUSSION THREAD

**Discussion by denniafredo** [singpass/demo-app#12] [Q&A]
Hi everyone,

We're currently integrating Singpass for authentication in our application, and we're encountering an issue during the login process. Every time we try to log in using the provided credentials (Singpass ID : F1612351W, Password : MyInfo2o15), it redirects us to the face verification step. This step is problematic for our team since not all of our developers are located in Singapore.

We are using an /auth endpoint in our application, which generates the authorization URL and redirects users to the Singpass login page. For testing and development purposes, is there a way to bypass or skip the face verification step? Alternatively, are there any sandbox or testing environments provided by Singpass that allow us to simulate this process without requiring actual face verification?

Your guidance or suggestions would be greatly appreciated. Thank you!

Here is my url : https://stg-id.singpass.gov.sg/auth?client_id=Dfn5pqzg4RwCpsJ28Fca83dcVHFVDKup&scope=email%20mobileno%20name%20openid%20regadd%20uinfin&response_type=code&redirect_uri=https%3A%2F%2Fspmw.interaktiv.sg%2Fapi%2Fcallback&code_challenge_method=S256&code_challenge=bxqeZ8NhZK5F2nCqbEOcb8sbn5zIck59Lis5pVZ_g7I&nonce=f67931e5-7d2d-4b47-9acd-1e305a24543d&state=9a2f750be346451925e691acce6dc728

![image](https://github.com/user-attachments/assets/7f6d6fd1-065a-42eb-b2dc-a58f08505982)

![image](https://github.com/user-attachments/assets/286e4419-527d-4bb1-8ef1-0a81d53eaa2b)

**Comment by luiscvega** [comment]
I tried on the mobile browser of Safari and I'm getting the same issue, it's asking for a Face Verification.

However, when I try on a desktop browser, I'm getting an authentication error. See the screenshot below:

![image](https://github.com/user-attachments/assets/d0733ae2-8d4b-4903-96ef-929d684c5880)

**Comment by dannycsz** [comment]
I have the same problem, I tried other test user from the persona list, it's either prompt for 2FA or Face Verification. Nothing works so far.

**Comment by poopypoops** [comment]
can you try using this test account - S7790795E password: Demo@123

**Reply by luiscvega** [reply]
Just tried it and still doesn't work.

**Reply by poopypoops** [reply]
use this password instead - T4!vY9@xL7$p

**Reply by luiscvega** [reply]
Hmm, just tried `S7790795E` with `T4!vY9@xL7$p` and also doesn't work.

**Comment by dannycsz** [comment]
My side working fine with this.

**Reply by denniafredo** [reply]
Hi @dannycsz , you could share the code or approach you use for the /auth endpoints?

I am currently working on a similar feature and would greatly appreciate any guidance or examples you could provide.
Thanks!

**Reply by dannycsz** [reply]
I mean I try the demo and it works but once I change the client to my staging app, it always return AUTH-E0001 even with the exact same keys.

**Comment by denniafredo** [comment]
![image](https://github.com/user-attachments/assets/db244243-5971-4e16-97f7-8f6712f59f8c)
i have this problem

**Comment by luiscvega** [comment]
@denniafredo @dannycsz @poopypoops

If you generate the `auth` URL correctly and see the sign-in page, it works when you enter `S7790795E` with `T4!vY9@xL7$p` in Chrome. But it does not work on Firefox. Strange. Using it on mobile also produces a different result. Highly inconsistent. Will test further.

**Comment by casuarinao8** [comment]
Anyone found a solution to this? I am also facing the same issue where System Code: AUTH-E0001b is shown.

**Comment by luiscvega** [comment]
I have been successful in retrieving the `/userinfo` so I'd be glad to assist anyone.

Here is the code that I used to try and successfully generate a working `/auth` URL:

```go
package main

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"fmt"
	"net/url"

	"github.com/google/uuid"
)

const (
	clientId    = "<your-client-id>"
	redirectUri = "http://localhost:8080/singpass"
	scopes      = "birthcountry dob email marital merdekagen.eligibility name nationality openid race sex uinfin"
	tokenUrl    = "https://stg-id.singpass.gov.sg/token"
)

func main() {
	// Step 1: Get authorize url
	_, codeChallenge, err := generatePkceCodePair()
	if err != nil {
		panic("failed to generate PKCE code pair!")
	}

	fmt.Println(authorizeUrl(codeChallenge))
}

func generatePkceCodePair() (string, string, error) {
	// code verifier
	bs := make([]byte, 32)

	if _, err := rand.Read(bs); err != nil {
		return "", "", err
	}

	codeVerifier := hex.EncodeToString(bs)

	// code challenge
	hash := sha256.New()

	if _, err := hash.Write([]byte(codeVerifier)); err != nil {
		return "", "", err
	}

	codeChallenge := base64.RawURLEncoding.EncodeToString(hash.Sum(nil))

	return codeVerifier, codeChallenge, nil
}

func authorizeUrl(codeChallenge string) string {
	v := url.Values{
		"scope":                 []string{scopes},
		"response_type":         []string{"code"},
		"redirect_uri":          []string{redirectUri},
		"nonce":                 []string{uuid.New().String()},
		"client_id":             []string{clientId},
		"state":                 []string{uuid.New().String()},
		"code_challenge":        []string{codeChallenge},
		"code_challenge_method": []string{"S256"},
	}

	return fmt.Sprintf("https://stg-id.singpass.gov.sg/auth?%s", v.Encode())
}
```

However, you need to make sure that your SDP settings are correct:

1. Make sure that your Redirect URL is correct and that you are using it in the code correctly. You'll notice in the code, my redirect URI is `http://localhost:8080/singpass`. This is also exactly is what is set in my SDP settings.
2. Your JWKS *must be valid*. If you're using the the JWKS URL, it must be publicly accessible and have a valid JWKS object. If you're using the JWKS Object text area, you must use a valid JWKS object as well. Here is a sample one from the `demo-app` (of course, you must generate your own valid ones when you are testing):

```json
{
  "keys": [
    {
      "kid": "my-sig-key",
      "kty": "EC",
      "use": "sig",
      "alg": "ES256",
      "crv": "P-256",
      "x": "16Ziaq7VMza-agHmM3P6uh887g2Yusp_lE7RBpi1YYI",
      "y": "3XidKG6ImEvtt9dcjX8UoXykPRTovfALxPXaWPBQ7yc"
    },
    {
      "kid": "my-enc-key",
      "kty": "EC",
      "use": "enc",
      "alg": "ECDH-ES+A256KW",
      "crv": "P-256",
      "x": "6MYhNwxiDqIlp2-x7-EXWx9lv6N1C7rqezs7Nl9j7hA",
      "y": "Q2cxMkolRlpdw8QY4Kwpkwqyp1vpoA3KVzoLXG2wvfQ"
    }
  ]
}
```

3. The only Myinfo Test Persona that works is this: `S7790795E` / `T4!vY9@xL7$p`. I've tried other Test Personas and they are not working. Probably an issue with Singpass.

3. Finally, I've noticed that the app only works in Google Chrome. In other browsers, I get the "We are sorry but there seems to be an error" page with the `AUTH-E0001` whenever I use the above persona. Again, I think this is an issue with Singpass.

**Reply by casuarinao8** [reply]
Could I understand whether u found any solution for this?

**Comment by luiscvega** [comment]
@denniafredo, are you still running into any  issues?

**Reply by denniafredo** [reply]
![image](https://github.com/user-attachments/assets/220892c4-1ab0-4bac-9751-5a2b334e5ac6)
yes, still the same issue, im using chrome btw

**Reply by denniafredo** [reply]
This is my node js code from demo-app, maybe you can give me some advice.

```
import config from './config.mjs';
import { Issuer, generators, custom } from 'openid-client';
import * as crypto from 'crypto';
import Router from '@koa/router';

// This demo uses panva/node-openid-client, an off-the-shelf OIDC client.

const singpassIssuer = await Issuer.discover(config.ISSUER_URL);

const singpassClient = new singpassIssuer.Client(
  {
    client_id: config.CLIENT_ID,
    response_types: ['code'],
    token_endpoint_auth_method: 'private_key_jwt',
    id_token_signed_response_alg: config.KEYS.PRIVATE_SIG_KEY.alg,
    userinfo_encrypted_response_alg: config.KEYS.PRIVATE_ENC_KEY.alg,
    userinfo_encrypted_response_enc: 'A256GCM',
    userinfo_signed_response_alg: config.KEYS.PRIVATE_SIG_KEY.alg,
  },
  { keys: [config.KEYS.PRIVATE_SIG_KEY, config.KEYS.PRIVATE_ENC_KEY] }
);

custom.setHttpOptionsDefaults({
  timeout: 15000,
});

const router = new Router();

router.get('/api/.well-known/jwks.json', function getJwks(ctx) {
  ctx.body = { keys: [config.KEYS.PUBLIC_SIG_KEY, config.KEYS.PUBLIC_ENC_KEY] };
});

router.get('/api/authData', async function handleAuth(ctx) {
  try {
    const code_verifier = generators.codeVerifier();
    const code_challenge = generators.codeChallenge(code_verifier);
    const nonce = crypto.randomUUID();
    const state = crypto.randomBytes(16).toString('hex');

    // Generate the authorization URL to redirect the user to the OIDC provider's authorization endpoint
    const authorizationUrl = singpassClient.authorizationUrl({
      redirect_uri: config.REDIRECT_URI,
      code_challenge_method: 'S256',
      code_challenge,
      nonce,
      state,
      scope: config.SCOPES,
    });

    ctx.body = {
      authorizationUrl,
      code_challenge,
      nonce,
      state
    };

  } catch (err) {
    console.error('Error initiating authentication:', err);
    ctx.status = 500;
    ctx.body = 'Authentication request failed';
  }
});

router.get('/api/callback', async function handleAuthCallback(ctx) {
  try {
    const {
      code_verifier,
      nonce,
      state,
    } = ctx.request.query;

    // Exchange the authorization code for tokens
    const tokenSet = await singpassClient.callback(config.REDIRECT_URI, receivedQueryParams, {
      code_verifier,
      nonce,
      state,
    });
    console.log('ID token claims:', tokenSet.claims());

    // Request user information if additional scopes are provided (e.g., 'profile', 'email')
    const userInfo = await singpassClient.userinfo(tokenSet.access_token);
    console.log('User info:', userInfo);

    // Store the user's session information (claims and user info)
    ctx.body = { ...tokenSet.claims(), ...userInfo };
  } catch (err) {
    console.error('Error handling authentication callback:', err);
    ctx.status = 401;
    ctx.body = 'Authentication failed';
  }
});



export { router };

```

**Reply by luiscvega** [reply]
Hi @denniafredo . This could be due to several factors. In order to help you, I'll need to know what your current setup is. Kindly see answer here to diagnose your issue:

https://github.com/singpass/demo-app/discussions/12#discussioncomment-10971576

Here are some questions:
1. Do you have a valid JWKS endpoint/object?
2. Are you generating your Auth URL correctly? Can you share an example of your URL?
3. What test persona are you using? This currently works: `Username: T4!vY9@xL7$p Password: T4!vY9@xL7$p`.
4. Is your redirect URL in your SDP settings exactly the same as the one when you generate your `/auth` URL? If you are using the demo-app settings, then you also need use the same redirect URL in your SDP as what the demo-app is using, which is: `http://localhost:3080/callback`.
6. What scopes are you using?
7. What browser are you using?

**Reply by denniafredo** [reply]
![image](https://github.com/user-attachments/assets/4e086b4f-55c5-41c9-861f-4b2eb6a73c81)


now it shows like this.

are you get this issue too @luiscvega ?

**Reply by luiscvega** [reply]
What if your “Authentication flow” setting? Can you try setting it to “1FA”?

**Reply by denniafredo** [reply]
@luiscvega can you assist me where to change the Authentication flow to 1FA ?

**Reply by luiscvega** [reply]
Sorry for the delayed response. You can change it in the SDP settings:

![image](https://github.com/user-attachments/assets/3e8ee02d-8048-4a94-960d-ca2b6d0267e5)

**Reply by denniafredo** [reply]
@luiscvega  Thank you for your help, as I don't have access to the dashboard.
